crates/codegen/parser/generator/src/parser_definition.rs

```diff
@@ -102,91 +105,71 @@ impl ParserDefinitionNodeExtensions for ParserDefinitionNode {
 
             Self::ScannerDefinition(scanner_definition, _) => {
                 let kind = format_ident!("{name}", name = scanner_definition.name());
-                if is_trivia {
-                    let function_name =
-                        format_ident!("{}_parse_token", context_name.to_snake_case());
-                    quote! {
-                        self.#function_name(input, TokenKind::#kind)
-                    }
+
+                let parse_token = if is_trivia {
+                    format_ident!("parse_token")
                 } else {
-                    let function_name =
-                        format_ident!("{}_parse_token_with_trivia", context_name.to_snake_case());
-                    quote! {
-                        self.#function_name(input, TokenKind::#kind)
-                    }
+                    format_ident!("parse_token_with_trivia")
+                };
+
+                quote! {
+                    self.#parse_token::<#lex_ctx>(input, TokenKind::#kind)
                 }
             }
 
             Self::TriviaParserDefinition(trivia_parser_definition, _) => {
-                let function_name = format_ident!(
-                    "{snake_case}",
-                    snake_case = trivia_parser_definition.name().to_snake_case()
-                );
+                let function_name =
+                    format_ident!("{}", trivia_parser_definition.name().to_snake_case());
+
                 quote! { self.#function_name(input) }
             }
 
             Self::ParserDefinition(parser_definition, _) => {
-                if is_trivia {
-                    unreachable!(
-                        "Trivia productions can only reference trivia or token productions"
-                    )
-                }
+                assert!(
```

Comment:
unlrelated to this PR:

I have a backlog/list of missing validation rules that I want to add sometime in the next few weeks when I get more time, rewriting/expanding the existing validation. Please feel free to let me know or file issues if you think there is anything that can be added, so that we are not forced to do these validations and possibly panic here.

Reply:
Sounds reasonable to verify this once; cc @AntonyBlakey who originally put this. We should make this as part of our migration to a single definition, probably